### DISCUSSION TITLE
How to get the dynamic equation?

### DISCUSSION THREAD

**Discussion by foolyc** [bulletphysics/bullet3#3746] [General]
![image](https://user-images.githubusercontent.com/7566074/35427274-4e97d9ce-02a3-11e8-9793-bb6bed09724a.png)

how to get the matrix  M/N/S/Jc etc. after loading a  robot xml file?

**Comment by erwincoumans** [comment]
As described in the [PyBullet Quickstart Guide](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.2ye70wns7io3), PyBullet provides a function to calculate the mass matrix M, calculateMassMatrix, and Jacobian, calculateJacobian. There is an example for calculateJacobian here:
https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/jacobian.py

There is also a [calculateInverseDynamics](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/inverse_dynamics.py) and a [calculateInverseKinematics](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/inverse_kinematics.py) with example:


Presumably the 'N' stands for Coriolis and centrifugal terms? Those are not exposed at the moment. Does 'S' stand for user forces? Some link/info about the variable names and what/why you need it would help.

**Comment by foolyc** [comment]
well, 'N' stands for centripetal, Coriolis and gravity forces, '\tao' stands for user force, '\lamda' stands for constraint forces
i want to use this matrix to do inverse dynamic QP optimization for control purpose.

**Comment by haudren** [comment]
Indeed, those terms are not exposed yet. I've had success using [RBDL](https://rbdl.bitbucket.io/) and [RBDyn](https://github.com/jrl-umi3218/RBDyn) (and maybe even [Tasks](https://github.com/jrl-umi3218/Tasks) if you want to do QP-based control) to compute the missing terms, treating pybullet as a pure simulator.

**Comment by erwincoumans** [comment]
It would be good to expose the missing terms in PyBullet as well.
1) Exposing the N term (gravity, coriolis and centrifugal) should be easy.

2) Why would you need the \tao / user force? Didn't you apply it yourself?

3) Do you need the constaint forces, as computed by the LCP solver, such as contact, joint limits term too?

How does RBDL or RBDyn help with (2) and (3)?

Any other term(s)?

**Comment by haudren** [comment]
Sorry, I did not come back to this issue:

RBDL/RBDyn only help with your point 1) AFAIK. However, RBDyn especially has a suite of associated control tools.

2) In my mind, `\tau` is the control signal, so it is provided by the user.

3) In typical control schemes, the contact forces are either sensed or estimated. I think it would be important to be able to retrieve that information from the simulator to implement and / or verify algorithms. An extra term might be the contact locations.

**Comment by elkmug** [comment]
>     1. Exposing the N term (gravity, coriolis and centrifugal) should be easy.

If we're voting, heck even if we're not, I'd love to see this exposed in pybullet.

**Comment by foolyc** [comment]
Thanks for your reply. I choose RBDL in my project.

**Comment by benelot** [comment]
Constraint forces might be something I need at some point for openai gym.

**Comment by erwincoumans** [comment]
Re-opened for others who may be interested:

I'm currently re-implementing DeepMimic in PyBullet and for stable PD control also need the N (or C, gravity) term. [DeepMimic](https://xbpeng.github.io/projects/DeepMimic/index.html) also uses Bullet, but it implements its own inverse dynamics and stable PD control, using those terms.
You should be able to get the N term using pybullet.calculateInverseDynamics and setting the target acceleration to zero vector:

N(q,qdot) = p.calculateInverseDynamics(objUid, q, qdot, zeroAcc)
I just added a [PR fix](https://github.com/bulletphysics/bullet3/pull/2007) to allow calculateInverseDynamics for floating bodies (it was only supporting fixed-base bodies with joints).

The contact forces are available too, after stepSimulation. Use pybullet.getContactPoints, and see the [quickstart guide](https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.cb0co8y2vuvc) for the applied force field.

The only remaining thing would be to expose the non-contact constraint forces (motors, limits etc) as well.

**Comment by elkmug** [comment]
This is cool.  My interest in the N term was because I wanted to experiment with stable PD control, as opposed to regular PD control, in PyBullet.  And, that interest was due to reading about them in the DeepMimic paper.

So I'm doubly happy.

**Comment by erwincoumans** [comment]
I'm currently looking to extend BulletInverseDynamics with spherical joints and add stable PD control, but it is takes some time.

DeepMimic Stable PD control is here:
https://github.com/xbpeng/DeepMimic/blob/master/DeepMimicCore/sim/ImpPDController.cpp#L136

Its inverse dynamics is here:
https://github.com/xbpeng/DeepMimic/blob/master/DeepMimicCore/sim/RBDUtil.cpp

**Comment by soshishimada** [comment]
In the presence of contact force from a floor, calculateInverseDynamics() function with zero acceleration will return:

N(q,qdot) - '\lamda' = p.calculateInverseDynamics(objUid, q, qdot, zeroAcc).

meaning, the returned value includes centripetal, Coriolis, gravity and force from a floor.  Am I right?
